Build failure: gitlab-container-registry (broken tests)

### Nixpkgs version

- Stable (24.11)

### Steps to reproduce

As already noted in #377301, the tests of `gitlab-container-registry-4.15.2 currently fail on NixOS-24.11.

But it's also failing as well on master (8da68c84cd092e76b8c4321dbaa22046c24725c2).

Built via `nix-build . -A gitlab-container-registry`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/287489183

### Relevant log output

```console
release-24.11:

buildPhase completed in 38 seconds
Running phase: checkPhase
ok      github.com/docker/distribution/configuration    0.041s
ok      github.com/docker/distribution/context  2.008s
ok      github.com/docker/distribution/health   0.006s
ok      github.com/docker/distribution/health/api       0.004s
ok      github.com/docker/distribution/health/checks    0.211s
ok      github.com/docker/distribution/internal/feature 0.004s
ok      github.com/docker/distribution/manifest/manifestlist    0.006s
ok      github.com/docker/distribution/manifest/manifestlist/compat     0.007s
ok      github.com/docker/distribution/manifest/ocischema       0.007s
ok      github.com/docker/distribution/manifest/schema1 0.009s
ok      github.com/docker/distribution/manifest/schema2 0.007s
ok      github.com/docker/distribution/notifications    3.586s
ok      github.com/docker/distribution/reference        0.009s
ok      github.com/docker/distribution/registry 28.141s
ok      github.com/docker/distribution/registry/api/errcode     0.005s
ok      github.com/docker/distribution/registry/api/urls        0.017s
ok      github.com/docker/distribution/registry/api/v2  0.298s
ok      github.com/docker/distribution/registry/auth/silly      0.006s
ok      github.com/docker/distribution/registry/auth/token      0.015s
ok      github.com/docker/distribution/registry/bbm     1.700s
ok      github.com/docker/distribution/registry/bbm/metrics     0.012s
ok      github.com/docker/distribution/registry/client  0.032s
ok      github.com/docker/distribution/registry/client/auth     0.021s
ok      github.com/docker/distribution/registry/client/auth/challenge   0.004s
ok      github.com/docker/distribution/registry/datastore       0.649s
ok      github.com/docker/distribution/registry/datastore/metrics       0.013s
?       github.com/docker/distribution/registry/datastore/migrations    [no test files]
ok      github.com/docker/distribution/registry/datastore/models        0.004s
ok      github.com/docker/distribution/registry/gc      0.014s
ok      github.com/docker/distribution/registry/gc/internal/metrics     0.014s
ok      github.com/docker/distribution/registry/gc/worker       0.065s
ok      github.com/docker/distribution/registry/handlers        36.087s
ok      github.com/docker/distribution/registry/internal/metrics/redis  0.008s
ok      github.com/docker/distribution/registry/internal/redis  0.006s
ok      github.com/docker/distribution/registry/storage 25.670s
ok      github.com/docker/distribution/registry/storage/cache   0.007s
ok      github.com/docker/distribution/registry/storage/cache/memory    0.008s
?       github.com/docker/distribution/registry/storage/cache/redis     [no test files]
ok      github.com/docker/distribution/registry/storage/driver  0.004s
--- FAIL: TestAzureDriver_parseParameters_Bool (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbf60d4]

goroutine 88 [running]:
testing.tRunner.func1.2({0xf02960, 0x193c120})
        /nix/store/2xc5jh4nxvm68c67ila9p2rj3qrjgz0x-go-1.23.5/share/go/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
        /nix/store/2xc5jh4nxvm68c67ila9p2rj3qrjgz0x-go-1.23.5/share/go/src/testing/testing.go:1635 +0x35e
panic({0xf02960?, 0x193c120?})
        /nix/store/2xc5jh4nxvm68c67ila9p2rj3qrjgz0x-go-1.23.5/share/go/src/runtime/panic.go:785 +0x132
github.com/docker/distribution/registry/storage/driver/internal/testutil.TestBoolValue(0xc00015cea0, {{0xea2ca0, 0x18efe48}, 0x0, 0x0, 0x0, 0x0, {0x107bbb1, 0x14}, {0x107bbed, ...}, ...})
        /build/source/registry/storage/driver/internal/testutil/testutil.go:41 +0x114
github.com/docker/distribution/registry/storage/driver/internal/testutil.AssertByDefaultType(0xc00015cea0, {{0xea2ca0, 0x18efe48}, 0x0, 0x0, 0x0, 0x0, {0x107bbb1, 0x14}, {0x107bbed, ...}, ...})
        /build/source/registry/storage/driver/internal/testutil/testutil.go:27 +0xfb
github.com/docker/distribution/registry/storage/driver/azure.TestAzureDriver_parseParameters_Bool(0xc00015cea0)
        /build/source/registry/storage/driver/azure/azure_test.go:321 +0x1bb
testing.tRunner(0xc00015cea0, 0x10cea98)
        /nix/store/2xc5jh4nxvm68c67ila9p2rj3qrjgz0x-go-1.23.5/share/go/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
        /nix/store/2xc5jh4nxvm68c67ila9p2rj3qrjgz0x-go-1.23.5/share/go/src/testing/testing.go:1743 +0x390
FAIL    github.com/docker/distribution/registry/storage/driver/azure    0.015s



master:

?       github.com/docker/distribution/registry/storage/cache/redis     [no test files]
ok      github.com/docker/distribution/registry/storage/driver  0.004s
--- FAIL: TestAzureDriver_parseParameters_Bool (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xbf61f4]

goroutine 90 [running]:
testing.tRunner.func1.2({0xf02940, 0x193c120})
        /nix/store/0b6vsy4fa4i4qpk1011hi6251nwdg5y8-go-1.23.4/share/go/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
        /nix/store/0b6vsy4fa4i4qpk1011hi6251nwdg5y8-go-1.23.4/share/go/src/testing/testing.go:1635 +0x35e
panic({0xf02940?, 0x193c120?})
        /nix/store/0b6vsy4fa4i4qpk1011hi6251nwdg5y8-go-1.23.4/share/go/src/runtime/panic.go:785 +0x132
github.com/docker/distribution/registry/storage/driver/internal/testutil.TestBoolValue(0xc000196b60, {{0xea2ce0, 0x18efe48}, 0x0, 0x0, 0x0, 0x0, {0x107bc0b, 0x14}, {0x107bc47, ...}, ...})
        /build/source/registry/storage/driver/internal/testutil/testutil.go:41 +0x114
github.com/docker/distribution/registry/storage/driver/internal/testutil.AssertByDefaultType(0xc000196b60, {{0xea2ce0, 0x18efe48}, 0x0, 0x0, 0x0, 0x0, {0x107bc0b, 0x14}, {0x107bc47, ...}, ...})
        /build/source/registry/storage/driver/internal/testutil/testutil.go:27 +0xfb
github.com/docker/distribution/registry/storage/driver/azure.TestAzureDriver_parseParameters_Bool(0xc000196b60)
        /build/source/registry/storage/driver/azure/azure_test.go:321 +0x1bb
testing.tRunner(0xc000196b60, 0x10ceaf8)
        /nix/store/0b6vsy4fa4i4qpk1011hi6251nwdg5y8-go-1.23.4/share/go/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
        /nix/store/0b6vsy4fa4i4qpk1011hi6251nwdg5y8-go-1.23.4/share/go/src/testing/testing.go:1743 +0x390
FAIL    github.com/docker/distribution/registry/storage/driver/azure    0.015s
FAIL
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.68, NixOS, 24.11 (Vicuna), 24.11.712148.edf04b75c13c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - channels(root): `"nixos-24.11.712148.edf04b75c13c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### Notify maintainers

@NixOS/gitlab @xanderio @blitz 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/gi/gitlab-container-registry/package.nix
+++ b/pkgs/by-name/gi/gitlab-container-registry/package.nix
@@ -2,6 +2,7 @@
   lib,
   buildGoModule,
   fetchFromGitLab,
+  fetchpatch,
 }:
 
 buildGoModule rec {
@@ -19,14 +20,35 @@ buildGoModule rec {
 
   vendorHash = "sha256-aKE/yr2Sh+4yw4TmpaVF84rJOI6cjs0DKY326+aXO1o=";
 
+  env = {
+    # required for multiple azure tests
+    # https://gitlab.com/gitlab-org/container-registry/-/issues/1494
+    AZURE_DRIVER_VERSION = "azure_v2";
+  };
+
+  patches = [
+    # remove with >= 4.15.3
+    (fetchpatch {
+      url = "https://gitlab.com/gitlab-org/container-registry/-/commit/268689a2f30880b7d122469a4260ca46cbc55ccd.patch";
+      hash = "sha256-RslK4qvcqCaG7ju2LgN/tI9cImrTj3Nry+mCv3zoWiA=";
+    })
+  ];
+
   postPatch = ''
     # Disable flaky inmemory storage driver test
     rm registry/storage/driver/inmemory/driver_test.go
 
     substituteInPlace health/checks/checks_test.go \
-      --replace \
+      --replace-fail \
         'func TestHTTPChecker(t *testing.T) {' \
         'func TestHTTPChecker(t *testing.T) { t.Skip("Test requires network connection")'
+
+    # Add workaround for failing test due to function type mismatch (args vs return) by upstream
+    # https://gitlab.com/gitlab-org/container-registry/-/issues/1495
+    substituteInPlace registry/storage/driver/base/regulator_test.go \
+      --replace-fail \
+        'require.Equal(t, limit, r.available, "r.available")' \
+        'require.Equal(t, limit, int(r.available), "r.available")'
   '';
 
   meta = with lib; {

